adam/visualization/positioning.py

```diff
@@ -602,5 +710,168 @@ def overlap_penalty(min_max_overlaps: torch.Tensor) -> torch.Tensor:
+            return torch.zeros(1)
+
+        if adam_object_2 in adam_object_1.relative_positions:
+            return RelationalPenalty.penalty(
+                bounding_box_1,
+                bounding_box_2,
+                adam_object_1.relative_positions[adam_object_2],
+            )
+
+    @staticmethod
+    def penalty(
+        box1: AxisAlignedBoundingBox,
+        box2: AxisAlignedBoundingBox,
+        box1_2_relation: InRegionDirectives,
+    ):
+        """
+        Assign a penalty for box1 if it does not comply with its relation to box2
+        Args:
+            box1:
+            box2:
+            box1_2_relation: relation of box1 to box2
+
+        Returns: Tensor(1,) with a penalty
+
+        """
+        # get direction that box 1 should be in w/r/t box 2
+        # TODO: allow for addressee directions
+        direction_vector = RelationalPenalty.direction_to_world(box1_2_relation.direction)
+
+        current_vector = box1.center - box2.center
+
+        angle = angle_between(direction_vector, current_vector)
+        if torch.isnan(angle):
+            angle = torch.zeros(1, dtype=torch.float)
+
+        # TODO: change to reflect distance from box extents
+        distance = box1.center_distance_from_point(box2.center)
+
+        # distal has a minimum distance away from object to qualify
+        if box1_2_relation.distance == DISTAL:
+            if distance < DISTAL_MIN_DISTANCE:
+                distance_penalty = DISTAL_MIN_DISTANCE - distance
+            else:
+                distance_penalty = torch.zeros(1)
+        # proximal has a min/max range
+        elif box1_2_relation.distance == PROXIMAL:
+            if PROXIMAL_MIN_DISTANCE <= distance <= PROXIMAL_MAX_DISTANCE:
+                distance_penalty = torch.zeros(1)
+            elif distance < PROXIMAL_MIN_DISTANCE:
+                distance_penalty = PROXIMAL_MIN_DISTANCE - distance
+            else:
+                distance_penalty = distance - PROXIMAL_MAX_DISTANCE
+
+        # exterior but in contact has a tiny epsilon of acceptable distance
+        # assuming that collisions are handled elsewhere
+        elif box1_2_relation.distance == EXTERIOR_BUT_IN_CONTACT:
+            if distance > EXTERIOR_BUT_IN_CONTACT_EPS:
+                distance_penalty = distance
+            else:
+                distance_penalty = torch.zeros(1)
+        else:
+            raise RuntimeError(
+                "Currently unable to support Interior distances w/ positioning solver"
+            )
+
+        print(
+            f"Angle penalty: {angle * ANGLE_PENALTY} + distance penalty: {distance_penalty * DISTANCE_PENALTY}"
+        )
+        return angle * ANGLE_PENALTY + distance_penalty * DISTANCE_PENALTY
+
+    @staticmethod
+    def direction_to_world(
+        direction: Direction[ReferenceObjectT],
+        direction_reference: Optional[AxisAlignedBoundingBox] = None,
```

Comment:
If we switch to working with `ObjectPerceptions`, then we just need to pass the `Direction` object plus our usual mapping of all `ObjectPerception`s to their bounding boxes.